### PR TITLE
[18807] win_firewall_rule module uses HNetCfg.FwPolicy2 COM Object

### LINKED ISSUE #18807: win_firewall_rule fails when locale isn't English
_From @ilpianista on November 29, 2016 10:45_

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
win_firewall_rule

##### ANSIBLE VERSION
```
ansible 2.2.0.0
```

##### OS / ENVIRONMENT
Target host is Windows 7 with German language.

##### SUMMARY
I've a task to open a firewall port.
The rule creation succeeded, but on the next execution of the same task it fails because win_firewall_rule tries to parse the output which is in German and not in English as the module does expect.

##### STEPS TO REPRODUCE
The rule is e.g.
```
name: Open required port on the firewall
win_firewall_rule: name="Foo 8080" localport="8080" protocol=TCP action=allow direction=In profile=public enable=yes state=present force=true
```
##### EXPECTED RESULTS
Task returns ok on every execution.

##### ACTUAL RESULTS
Task fails on rule check.

_Copied from original issue: ansible/ansible-modules-extras#3563_

### PR DESCRIPTION
**Pull request by ar7z1**
##### SUMMARY

In this PR I fixed #18807. This issue was closed by mistake, it's easily reproduces on my Russian Windows server 2008.

`netsh advfirewall firewall show` on non-English Windows produces non-English output:
```
PS C:\Users\Администратор> netsh advfirewall firewall show rule name=http

Имя правила:                          http
----------------------------------------------------------------------
Включен:                              Да
Направление:                          Вход
Profiles:                             Домен,Частный,Публичный
Группировка:
LocalIP:                              Любой
Удаленный IP-адрес:                   192.168.0.1/32
Протокол:                             TCP
Локальный порт:                       80
Удаленный порт:                       Любой
Обход узлов:                          Нет
Действие:                             Разрешить
ОК.
```
And there is no way to ask `netsh advfirewall firewall` command to produce output in English. :-(

In this PR I use `HNetCfg.FwPolicy2` COM object instead of `netsh advfirewall firewall` command.
It's supported from Windows Vista ([MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365309(v=vs.85).aspx)).

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

windows/win_firewall_rule

##### ANSIBLE VERSION

```
ansible 2.4.0 (18807-win_firewall_rule-for-non-english-locale 64b006c0ae) last updated 2017/07/27 16:19:56 (GMT +300)
  config file = /home/art/.ansible.cfg
  configured module search path = [u'/home/art/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/art/projects/github/ansible/lib/ansible
  executable location = /home/art/projects/github/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

```
```

### CODE REVIEW

**Comment by ar7z1** [comment]
In this PR I completely rewrote all module (and removed a lot of code). @dagwieers, can I ask you to review changes and give me feedback?

The PR has WIP status because I didn't add support for `edge`, `interfacetypes` and `security` properties.  And I'm going to add several new tests. But the main part of the work is already finished.

**Comment by ar7z1** [comment]
@ilpianista, can you help to test this PR?

**Comment by ar7z1** [comment]
I think that #25157 and #22977 also will be fixed in this PR.

**Comment by dagwieers** [comment]
@ar7z1 Whoaw, nice work. I remember we discussed this before and the consensus was to make use of Powershell cmdlets for the future implementation (hence the note in the source-code), I noticed you decided to use the COM object. I guess because it supports all versions of Windows.

@nitzmahone @jborean93 @jhawkesworth What is your opinion on this ?

**Comment by nitzmahone** [comment]
Fantastic- I *hate* wrapping legacy .exe's for modules, so this is a huge improvement. I'll be curious to see if they included those objects on Nano Server for managing the firewall or if we'll need a cmdlet codepath in there for that, but we'll jump off that bridge when we get to it.

**Comment by ar7z1** [comment]
@dagwieers, thank you for review!

> I noticed you decided to use the COM object. I guess because it supports all versions of Windows.

You're absolutely right! At first I was going to use [Get-NetFirewallRule](https://technet.microsoft.com/ru-ru/library/jj554860(v=wps.630).aspx) cmdlet, but found that it does not exist on my Windows Server 2008 R2 (it appeared in Windows 8.1 and Windows Server 2012 R2). :-(
So the only option was to use the `HNetCfg.FwPolicy2` COM object (it [appeared](https://technet.microsoft.com/en-us/library/2008.06.security.aspx) in Windows Vista).

**Comment by ar7z1** [comment]
@nitzmahone, thank you for review!

>I'll be curious to see if they included those objects on Nano Server for managing the firewall or if we'll need a cmdlet codepath in there for that, but we'll jump off that bridge when we get to it.

I tried to create `HNetCfg.FwPolicy2` COM object on Nano Server and it's successfully created! :-)
I couldn't run tests for win_firewall_rule module because of #23990, but I think all should work on Nano Server in the same way. I'll return to testing after fixing TODOs in this PR.

I plan to check that everything works on:
* Windows Vista??
* Windows 7
* Windows 8.1
* Windows 10
* Windows 2008 (Windows Server 2008 R2 already tested)
* Windows Server 2012
* Windows Server 2012 R2
* Windows Server 2016
* Windows Nano Server

**Comment by ilpianista** [comment]
@ar7z1 great work indeed! I can help testing on 7, 2008 R2 and 2012.

**Comment by ar7z1** [comment]
@ilpianista 
> I can help testing on 7, 2008 R2 and 2012.

Great news! I've just finished the PR. ;-)

File: test/integration/targets/win_firewall_rule/tasks/main.yml
Comment (ar7z1) [inline test/integration/targets/win_firewall_rule/tasks/main.yml]: Can't assert because test will fail on Windows < Windows 7/Windows Server 2008 R2.

File: test/integration/targets/win_firewall_rule/tasks/main.yml
Comment (ar7z1) [inline test/integration/targets/win_firewall_rule/tasks/main.yml]: Can't assert because test will fail on Windows < Windows 8/Windows Server 2012.

**Comment by ar7z1** [comment]
`ansible-test windows-integration -v win_firewall_rule` passed on:
* Windows 7
* Windows Server 2008 R2 SP1
* Windows 8.1
* Windows Server 2012 R2
* Windows 10
* Windows Server 2016

TODO test:
* Windows 2008
* Windows 8
* Windows Server 2012

File: test/integration/targets/win_firewall_rule/tasks/main.yml
Comment (jborean93) [reply test/integration/targets/win_firewall_rule/tasks/main.yml]: You can always add a conditional to the task to only run on particular hosts. In https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_share/tasks/main.yml we check if a particular cmdlet is available and only run subsequent tasks if it is, in your case you can either use the OS version or some other check before running this test.

File: test/integration/targets/win_firewall_rule/tasks/main.yml
Comment (jborean93) [reply test/integration/targets/win_firewall_rule/tasks/main.yml]: Ditto above

**Comment by jborean93** [comment]
@ar7z1, what was left to do on this, I'm able to test this sometime soon but didn't want to test all of it in case your are still making changes.

**Comment by dagwieers** [comment]
@ar7z1 And please add yourself to the list of authors too !

**Comment by ar7z1** [comment]
@jborean93 
> what was left to do on this, I'm able to test this sometime soon but didn't want to test all of it in case your are still making changes.

I finished writing code and now run tests for all windows versions what I have.
I have left to run tests on:
* Windows 7
* Windows 2008
* Windows 8
* Windows Server 2012
* Windows Nano Server

Right now I'm preparing Windows 7 box. :-)
And it will be cool if you also test this PR! ;-)

File: test/integration/targets/win_firewall_rule/tasks/main.yml
Comment (ar7z1) [reply test/integration/targets/win_firewall_rule/tasks/main.yml]: @jborean93 Thank you very much! I will add conditions. :-)

**Comment by jborean93** [comment]
Thanks, will do. I've got an environment with 2008 and 2012 so I will be able to test that. I'm not sure how you will go with Nano Server as I've never even connected Ansible to it yet.

File: test/integration/targets/win_firewall_rule/tasks/main.yml
Comment (ar7z1) [reply test/integration/targets/win_firewall_rule/tasks/main.yml]: @jborean93 [Added](https://github.com/ansible/ansible/pull/27381/commits/237b0d1f5188c2abe976132f9bbac06df93de64b) additional asserts. Thank you again!

**Comment by ar7z1** [comment]
@jborean93 
> I'm not sure how you will go with Nano Server as I've never even connected Ansible to it yet.

Well, then I think that I have to skip testing this PR on it. :-)

**Comment by ar7z1** [comment]
@dagwieers 
> And please add yourself to the list of authors too !

Should I just add my name to the header of module?

**Comment by ar7z1** [comment]
I successfully ran `ansible-test windows-integration -v win_firewall_rule` on:
* Windows 7
* Windows 2008 SP2
* Windows Server 2008 R2
* Windows 8.1
* Windows Server 2012 R2
* Windows 10
* Windows Server 2016

I didn't have distributives for Windows 8 and Windows Server 2012. Thats why I didn't test on them. :-(

@dagwieers, @nitzmahone, @jborean93 I think that this PR is ready for review.
@ilpianista, @jborean93 It will be awesome if you can find time to test this PR in your environments! :-)

**Review by dagwieers — APPROVED** [review]

**Comment by dagwieers** [comment]
I think this is ready to be shipped in v2.4. It includes integration tests, was tested against all supported Windows versions and the new maintainer is a great guy :-)

shipit

**Comment by dagwieers** [comment]
@ar7z1 You can add yourself as an author to the module and claim copyright too, since you basically rewrote the module. Or you can add yourself as a maintainer in github/BOTMETA.yml. But the former has the same effect.

**Comment by ar7z1** [comment]
@dagwieers Thank you for review! :-)
I [fixed](https://github.com/ansible/ansible/pull/27381/commits/23844194b8b6a909f0fef063487986c38a597f90) docs and copyrights.

**Review by dagwieers — APPROVED** [review]
LGTM

**Comment by ar7z1** [comment]
@dagwieers Thank you for review!

**Comment by ar7z1** [comment]
bot_status

**Comment by ar7z1** [comment]
Is the bot waiting for "shipit" from the community? @jborean93, can you review the changes?

**Comment by jborean93** [comment]
Sorry @ar7z1, I'm about to head off on a flight which will be over 24 hours. I'll have a look at it when I land.

**Comment by ar7z1** [comment]
@jborean93 Awesome! Thank you!

File: lib/ansible/modules/windows/win_firewall_rule.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_firewall_rule.ps1:186]: If `$force` isn't required anymore I would add a `Add-Warning` call to say it is no longer required.

File: lib/ansible/modules/windows/win_firewall_rule.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_firewall_rule.ps1:232]: What is the error message like if this goes wrong, should we be catching it and calling `Fail-Json` with a better error message instead?

**Review by jborean93 — COMMENTED** [review]
Looks like some good changes here, have left just 2 comments in line. Only other concern is that the return values from before are no longer being set. While they are not documented and this is a preview version, can we derive the same return values without resorting to cmd?

File: lib/ansible/modules/windows/win_firewall_rule.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_firewall_rule.ps1:53]: Instead of splitting the string here, would be better to change the type of `$profiles` with `Get-AnsibleParam` to list. It will automatically convert strings with `,` separators but also supports people passing in a list themselves.

File: lib/ansible/modules/windows/win_firewall_rule.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_firewall_rule.ps1:181]: See my point in `$profiles` regarding it being a list type and the same could be applied here.

**Comment by ar7z1** [comment]
@jborean93 Thank you for review! I'll try to fix all today. :-)

**Comment by jborean93** [comment]
@ar7z1, we've held this up for too long and the freeze is impending. I'm going to merge this through but feel free to raise another PR for some of the stuff above.

**Review by jborean93 — APPROVED** [review]

**Pull request merged**
